File: java/23. Merge k Sorted Lists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> heap = new PriorityQueue<>((a, b) -> a.val - b.val);

        for (ListNode list : lists) {
            if (list != null) heap.add(list);
        }

        ListNode preHead = new ListNode();
        ListNode current = preHead;
        while (!heap.isEmpty()) {
            ListNode node = heap.poll();
            if (node.next != null) heap.add(node.next);
            node.next = null;
            current.next = node;
            current = current.next;
        }

        return preHead.next;
    }

}

File: python/23. Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        def mergeTwoLists(left, right):
            preHead = ListNode()
            current = preHead

            while left and right:
                if left.val < right.val:
                    current.next = left
                    left = left.next
                else:
                    current.next = right
                    right = right.next
                current = current.next

            if not left:
                current.next = right
            if not right:
                current.next = left

            return preHead.next


        if not lists:
            return None
        if len(lists) == 1:
            return lists[0]

        while len(lists) > 1:
            lists.append(mergeTwoLists(lists.pop(), lists.pop()))

        return lists[0]